File: sdk/typescript/src/utils/request-signer.ts
/**
 * @fileoverview Request signing utilities for secure API authentication
 * @author SwappiQ Protocol
 * @description Production-grade request signing with multiple algorithms and security features
 */

import { createHmac, createHash, randomBytes, timingSafeEqual } from 'crypto';
import { AuthCredentials, SignedRequest } from '../types/api.js';

export interface SigningOptions {
  algorithm?: 'HMAC-SHA256' | 'HMAC-SHA512' | 'Ed25519';
  includeBody?: boolean;
  includeHeaders?: string[];
  timestampTolerance?: number;
  nonce?: boolean;
}

export interface SignatureComponents {
  method: string;
  path: string;
  timestamp: string;
  nonce?: string;
  body?: string;
  headers?: string;
}

/**
 * Enterprise-grade request signer with multiple security algorithms
 */
export class RequestSigner {
  private readonly credentials?: AuthCredentials;
  private readonly defaultOptions: Required<SigningOptions>;

  constructor(credentials?: AuthCredentials, options: SigningOptions = {}) {
    this.credentials = credentials;
    this.defaultOptions = {
      algorithm: 'HMAC-SHA256',
      includeBody: true,
      includeHeaders: ['content-type', 'x-timestamp'],
      timestampTolerance: 30000, // 30 seconds
      nonce: true,
      ...options
    };
  }

  /**
   * Sign an HTTP request for API authentication
   */
  async signRequest(
    request: Omit<SignedRequest, 'signature' | 'headers'>,
    options: SigningOptions = {}
  ): Promise<SignedRequest> {
    if (!this.credentials) {
      throw new Error('Authentication credentials not provided');
    }

    const opts = { ...this.defaultOptions, ...options };
    const timestamp = request.timestamp || Date.now().toString();
    const nonce = opts.nonce ? this.generateNonce() : undefined;

    // Build signature components
    const components = this.buildSignatureComponents({
      method: request.method,
      path: request.path,
      timestamp,
      nonce,
      body: opts.includeBody ? request.body : undefined
    }, opts);

    // Generate signature
    const signature = await this.generateSignature(components, opts.algorithm);

    // Build headers
    const headers = this.buildAuthHeaders({
      apiKey: this.credentials.apiKey,
      timestamp,
      nonce,
      signature,
      passphrase: this.credentials.passphrase
    });

    return {
      ...request,
      timestamp,
      signature,
      headers
    };
  }

  /**
   * Verify a signed request (for webhook validation)
   */
  async verifyRequest(
    request: SignedRequest,
    secret: string,
    options: SigningOptions = {}
  ): Promise<boolean> {
    try {
      const opts = { ...this.defaultOptions, ...options };
      
      // Check timestamp tolerance
      if (opts.timestampTolerance > 0) {
        const requestTime = parseInt(request.timestamp);
        const currentTime = Date.now();
        const timeDiff = Math.abs(currentTime - requestTime);
        
        if (timeDiff > opts.timestampTolerance) {
          return false;
        }
      }

      // Reconstruct signature components
      const components = this.buildSignatureComponents({
        method: request.method,
        path: request.path,
        timestamp: request.timestamp,
        nonce: this.extractNonceFromHeaders(request.headers),
        body: opts.includeBody ? request.body : undefined
      }, opts);

      // Generate expected signature
      const expectedSignature = await this.generateSignatureWithSecret(
        components,
        secret,
        opts.algorithm
      );

      // Timing-safe comparison
      return this.compareSignatures(request.signature, expectedSignature);

    } catch (error) {
      console.error('Request verification failed:', error);
      return false;
    }
  }

  /**
   * Generate signature from components
   */
  private async generateSignature(
    components: SignatureComponents,
    algorithm: string
  ): Promise<string> {
    if (!this.credentials?.apiSecret) {
      throw new Error('API secret not provided');
    }

    return this.generateSignatureWithSecret(components, this.credentials.apiSecret, algorithm);
  }

  /**
   * Generate signature with provided secret
   */
  private async generateSignatureWithSecret(
    components: SignatureComponents,
    secret: string,
    algorithm: string
  ): Promise<string> {
    const message = this.buildSignatureString(components);

    switch (algorithm) {
      case 'HMAC-SHA256':
        return createHmac('sha256', secret)
          .update(message)
          .digest('hex');

      case 'HMAC-SHA512':
        return createHmac('sha512', secret)
          .update(message)
          .digest('hex');

      case 'Ed25519':
        // For Ed25519, we'd typically use a specialized crypto library
        // This is a simplified implementation using HMAC-SHA256 as fallback
        console.warn('Ed25519 not fully implemented, falling back to HMAC-SHA256');
        return createHmac('sha256', secret)
          .update(message)
          .digest('hex');

      default:
        throw new Error(`Unsupported signing algorithm: ${algorithm}`);
    }
  }

  /**
   * Build signature components for message construction
   */
  private buildSignatureComponents(
    base: Partial<SignatureComponents>,
    options: SigningOptions
  ): SignatureComponents {
    const components: SignatureComponents = {
      method: base.method || '',
      path: base.path || '',
      timestamp: base.timestamp || Date.now().toString()
    };

    if (base.nonce) {
      components.nonce = base.nonce;
    }

    if (options.includeBody && base.body !== undefined) {
      components.body = this.normalizeBody(base.body);
    }

    if (options.includeHeaders && options.includeHeaders.length > 0) {
      components.headers = this.buildHeadersString(base.headers || {}, options.includeHeaders);
    }

    return components;
  }

  /**
   * Build the string to be signed
   */
  private buildSignatureString(components: SignatureComponents): string {
    const parts: string[] = [
      components.method.toUpperCase(),
      components.path,
      components.timestamp
    ];

    if (components.nonce) {
      parts.push(components.nonce);
    }

    if (components.body !== undefined) {
      parts.push(components.body);
    }

    if (components.headers) {
      parts.push(components.headers);
    }

    return parts.join('\n');
  }

  /**
   * Build authentication headers
   */
  private buildAuthHeaders(auth: {
    apiKey: string;
    timestamp: string;
    nonce?: string;
    signature: string;
    passphrase?: string;
  }): Record<string, string> {
    const headers: Record<string, string> = {
      'X-API-Key': auth.apiKey,
      'X-Timestamp': auth.timestamp,
      'X-Signature': auth.signature
    };

    if (auth.nonce) {
      headers['X-Nonce'] = auth.nonce;
    }

    if (auth.passphrase) {
      headers['X-Passphrase'] = auth.passphrase;
    }

    return headers;
  }

  /**
   * Generate cryptographically secure nonce
   */
  private generateNonce(length: number = 16): string {
    return randomBytes(length).toString('hex');
  }

  /**
   * Normalize request body for consistent signing
   */
  private normalizeBody(body: string): string {
    if (!body) return '';
    
    try {
      // Parse and re-stringify JSON to normalize formatting
      const parsed = JSON.parse(body);
      return JSON.stringify(parsed, Object.keys(parsed).sort());
    } catch {
      // If not JSON, return as-is
      return body;
    }
  }

  /**
   * Build headers string for signing
   */
  private buildHeadersString(
    headers: Record<string, string>,
    includeHeaders: string[]
  ): string {
    const normalizedHeaders: Record<string, string> = {};
    
    // Normalize header names to lowercase
    for (const [key, value] of Object.entries(headers)) {
      normalizedHeaders[key.toLowerCase()] = value;
    }

    // Build signed headers string
    const signedHeaders = includeHeaders
      .map(header => header.toLowerCase())
      .sort()
      .map(header => `${header}:${normalizedHeaders[header] || ''}`)
      .join('\n');

    return signedHeaders;
  }

  /**
   * Extract nonce from request headers
   */
  private extractNonceFromHeaders(headers: Record<string, string>): string | undefined {
    for (const [key, value] of Object.entries(headers)) {
      if (key.toLowerCase() === 'x-nonce') {
        return value;
      }
    }
    return undefined;
  }

  /**
   * Timing-safe signature comparison with enhanced security
   */
  private compareSignatures(signature1: string, signature2: string): boolean {
    // Always ensure equal length comparison for timing safety
    const len1 = signature1.length;
    const len2 = signature2.length;
    const maxLen = Math.max(len1, len2);
    
    // Pad shorter string with zeros to ensure constant-time comparison
    const padded1 = signature1.padEnd(maxLen, '0');
    const padded2 = signature2.padEnd(maxLen, '0');

    try {
      // Convert to buffers for timing-safe comparison
      const buffer1 = Buffer.from(padded1, 'hex');
      const buffer2 = Buffer.from(padded2, 'hex');
      
      // Additional length check after conversion
      if (buffer1.length !== buffer2.length) {
        return false;
      }
      
      // Use timing-safe comparison and verify original lengths match
      return timingSafeEqual(buffer1, buffer2) && len1 === len2;
    } catch (error) {
      // Secure fallback - no timing attack vulnerability
      let result = 0;
      
      // Always compare full padded length for timing safety
      for (let i = 0; i < maxLen; i++) {
        const char1 = i < len1 ? padded1.charCodeAt(i) : 0;
        const char2 = i < len2 ? padded2.charCodeAt(i) : 0;
        result |= char1 ^ char2;
      }
      
      // Only return true if both result is 0 AND lengths match
      return result === 0 && len1 === len2;
    }
  }

  /**
   * Create signature for webhook payload
   */
  createWebhookSignature(payload: string, secret: string, algorithm = 'sha256'): string {
    return createHmac(algorithm, secret)
      .update(payload)
      .digest('hex');
  }

  /**
   * Verify webhook signature
   */
  verifyWebhookSignature(
    payload: string,
    signature: string,
    secret: string,
    algorithm = 'sha256'
  ): boolean {
    const expectedSignature = this.createWebhookSignature(payload, secret, algorithm);
    return this.compareSignatures(signature, expectedSignature);
  }

  /**
   * Generate API key pair for development/testing
   */
  static generateApiKeyPair(): { apiKey: string; apiSecret: string } {
    const apiKey = 'sk_' + randomBytes(16).toString('hex');
    const apiSecret = randomBytes(32).toString('hex');
    
    return { apiKey, apiSecret };
  }

  /**
   * Hash sensitive data for logging
   */
  static hashForLogging(data: string): string {
    return createHash('sha256').update(data).digest('hex').substring(0, 8);
  }

  /**
   * Validate API key format
   */
  static validateApiKey(apiKey: string): boolean {
    // API keys should start with 'sk_' and be 35 characters total
    return /^sk_[a-f0-9]{32}$/.test(apiKey);
  }

  /**
   * Validate API secret format
   */
  static validateApiSecret(apiSecret: string): boolean {
    // API secrets should be 64 hex characters
    return /^[a-f0-9]{64}$/.test(apiSecret);
  }
}

/**
 * Utility function to create a configured request signer
 */
export function createRequestSigner(
  credentials: AuthCredentials,
  options?: SigningOptions
): RequestSigner {
  return new RequestSigner(credentials, options);
}

/**
 * Utility function for quick request signing
 */
export async function signRequest(
  request: Omit<SignedRequest, 'signature' | 'headers'>,
  credentials: AuthCredentials,
  options?: SigningOptions
): Promise<SignedRequest> {
  const signer = new RequestSigner(credentials, options);
  return signer.signRequest(request, options);
}

/**
 * Utility function for webhook signature verification
 */
export function verifyWebhookSignature(
  payload: string,
  signature: string,
  secret: string,
  algorithm = 'sha256'
): boolean {
  const signer = new RequestSigner();
  return signer.verifyWebhookSignature(payload, signature, secret, algorithm);
}

File: sdk/python/swappiq_sdk/request_signer.py
"""
Request signing utilities for secure API authentication
Author: SwappiQ Protocol
Description: Production-grade request signing with multiple algorithms and security features
"""

import hmac
import hashlib
import time
import json
import secrets
from typing import Dict, List, Optional, Literal
from dataclasses import dataclass
import base64

from .types import AuthCredentials, SignedRequest

@dataclass
class SigningOptions:
    algorithm: Literal['HMAC-SHA256', 'HMAC-SHA512'] = 'HMAC-SHA256'
    include_body: bool = True
    include_headers: List[str] = None
    timestamp_tolerance: int = 30000  # 30 seconds
    nonce: bool = True

@dataclass
class SignatureComponents:
    method: str
    path: str
    timestamp: str
    nonce: Optional[str] = None
    body: Optional[str] = None
    headers: Optional[str] = None

class RequestSigner:
    """Enterprise-grade request signer with multiple security algorithms"""
    
    def __init__(self, credentials: Optional[AuthCredentials], options: SigningOptions = None):
        self.credentials = credentials
        self.default_options = options or SigningOptions()
        if self.default_options.include_headers is None:
            self.default_options.include_headers = ['content-type', 'x-timestamp']
    
    async def sign_request(
        self,
        method: str,
        path: str,
        body: str = '',
        timestamp: Optional[str] = None,
        options: Optional[SigningOptions] = None
    ) -> SignedRequest:
        """Sign an HTTP request for API authentication"""
        if not self.credentials:
            raise ValueError("Authentication credentials not provided")
            
        opts = options or self.default_options
        timestamp = timestamp or str(int(time.time() * 1000))
        nonce = self._generate_nonce() if opts.nonce else None
        
        # Build signature components
        components = self._build_signature_components(
            method=method,
            path=path,
            timestamp=timestamp,
            nonce=nonce,
            body=body if opts.include_body else None
        )
        
        # Generate signature
        signature = await self._generate_signature(components, opts.algorithm)
        
        # Build headers
        headers = self._build_auth_headers(
            api_key=self.credentials.api_key,
            timestamp=timestamp,
            nonce=nonce,
            signature=signature,
            passphrase=self.credentials.passphrase
        )
        
        return SignedRequest(
            method=method,
            path=path,
            body=body,
            timestamp=timestamp,
            signature=signature,
            headers=headers
        )
    
    async def verify_request(
        self,
        request: SignedRequest,
        secret: str,
        options: Optional[SigningOptions] = None
    ) -> bool:
        """Verify a signed request (for webhook validation)"""
        try:
            opts = options or self.default_options
            
            # Check timestamp tolerance
            if opts.timestamp_tolerance > 0:
                request_time = int(request.timestamp)
                current_time = int(time.time() * 1000)
                time_diff = abs(current_time - request_time)
                
                if time_diff > opts.timestamp_tolerance:
                    return False
            
            # Reconstruct signature components
            nonce = self._extract_nonce_from_headers(request.headers)
            components = self._build_signature_components(
                method=request.method,
                path=request.path,
                timestamp=request.timestamp,
                nonce=nonce,
                body=request.body if opts.include_body else None
            )
            
            # Generate expected signature
            expected_signature = await self._generate_signature_with_secret(
                components, secret, opts.algorithm
            )
            
            # Timing-safe comparison
            return self._compare_signatures(request.signature, expected_signature)
            
        except Exception as e:
            print(f"Request verification failed: {e}")
            return False
    
    async def _generate_signature(self, components: SignatureComponents, algorithm: str) -> str:
        """Generate signature from components"""
        if not self.credentials or not self.credentials.api_secret:
            raise ValueError("API secret not provided")
            
        return await self._generate_signature_with_secret(
            components, self.credentials.api_secret, algorithm
        )
    
    async def _generate_signature_with_secret(
        self, components: SignatureComponents, secret: str, algorithm: str
    ) -> str:
        """Generate signature with provided secret"""
        message = self._build_signature_string(components)
        
        if algorithm == 'HMAC-SHA256':
            return hmac.new(
                secret.encode('utf-8'),
                message.encode('utf-8'),
                hashlib.sha256
            ).hexdigest()
        elif algorithm == 'HMAC-SHA512':
            return hmac.new(
                secret.encode('utf-8'),
                message.encode('utf-8'),
                hashlib.sha512
            ).hexdigest()
        else:
            raise ValueError(f"Unsupported signing algorithm: {algorithm}")
    
    def _build_signature_components(
        self,
        method: str,
        path: str,
        timestamp: str,
        nonce: Optional[str] = None,
        body: Optional[str] = None
    ) -> SignatureComponents:
        """Build signature components for message construction"""
        return SignatureComponents(
            method=method.upper(),
            path=path,
            timestamp=timestamp,
            nonce=nonce,
            body=self._normalize_body(body) if body else None
        )
    
    def _build_signature_string(self, components: SignatureComponents) -> str:
        """Build the string to be signed"""
        parts = [
            components.method,
            components.path,
            components.timestamp
        ]
        
        if components.nonce:
            parts.append(components.nonce)
            
        if components.body is not None:
            parts.append(components.body)
            
        if components.headers:
            parts.append(components.headers)
            
        return '\n'.join(parts)
    
    def _build_auth_headers(
        self,
        api_key: str,
        timestamp: str,
        nonce: Optional[str],
        signature: str,
        passphrase: Optional[str]
    ) -> Dict[str, str]:
        """Build authentication headers"""
        headers = {
            'X-API-Key': api_key,
            'X-Timestamp': timestamp,
            'X-Signature': signature
        }
        
        if nonce:
            headers['X-Nonce'] = nonce
            
        if passphrase:
            headers['X-Passphrase'] = passphrase
            
        return headers
    
    def _generate_nonce(self, length: int = 16) -> str:
        """Generate cryptographically secure nonce"""
        return secrets.token_hex(length)
    
    def _normalize_body(self, body: str) -> str:
        """Normalize request body for consistent signing"""
        if not body:
            return ''
            
        try:
            # Parse and re-stringify JSON to normalize formatting
            parsed = json.loads(body)
            return json.dumps(parsed, sort_keys=True, separators=(',', ':'))
        except json.JSONDecodeError:
            # If not JSON, return as-is
            return body
    
    def _extract_nonce_from_headers(self, headers: Dict[str, str]) -> Optional[str]:
        """Extract nonce from request headers"""
        for key, value in headers.items():
            if key.lower() == 'x-nonce':
                return value
        return None
    
    def _compare_signatures(self, signature1: str, signature2: str) -> bool:
        """Timing-safe signature comparison"""
        if len(signature1) != len(signature2):
            return False
            
        return hmac.compare_digest(signature1.encode(), signature2.encode())
    
    def create_webhook_signature(self, payload: str, secret: str, algorithm: str = 'sha256') -> str:
        """Create signature for webhook payload"""
        return hmac.new(
            secret.encode('utf-8'),
            payload.encode('utf-8'),
            getattr(hashlib, algorithm)
        ).hexdigest()
    
    def verify_webhook_signature(
        self, payload: str, signature: str, secret: str, algorithm: str = 'sha256'
    ) -> bool:
        """Verify webhook signature"""
        expected_signature = self.create_webhook_signature(payload, secret, algorithm)
        return self._compare_signatures(signature, expected_signature)
    
    @staticmethod
    def generate_api_key_pair() -> Dict[str, str]:
        """Generate API key pair for development/testing"""
        api_key = 'sk_' + secrets.token_hex(16)
        api_secret = secrets.token_hex(32)
        
        return {'api_key': api_key, 'api_secret': api_secret}
    
    @staticmethod
    def hash_for_logging(data: str) -> str:
        """Hash sensitive data for logging"""
        return hashlib.sha256(data.encode()).hexdigest()[:8]
    
    @staticmethod
    def validate_api_key(api_key: str) -> bool:
        """Validate API key format"""
        import re
        return bool(re.match(r'^sk_[a-f0-9]{32}$', api_key))
    
    @staticmethod
    def validate_api_secret(api_secret: str) -> bool:
        """Validate API secret format"""
        import re
        return bool(re.match(r'^[a-f0-9]{64}$', api_secret))

# Utility functions
def create_request_signer(
    credentials: AuthCredentials, options: Optional[SigningOptions] = None
) -> RequestSigner:
    """Create a configured request signer"""
    return RequestSigner(credentials, options)

async def sign_request(
    method: str,
    path: str,
    body: str,
    credentials: AuthCredentials,
    options: Optional[SigningOptions] = None
) -> SignedRequest:
    """Utility function for quick request signing"""
    signer = RequestSigner(credentials, options)
    return await signer.sign_request(method, path, body)

def verify_webhook_signature(
    payload: str, signature: str, secret: str, algorithm: str = 'sha256'
) -> bool:
    """Utility function for webhook signature verification"""
    signer = RequestSigner(None)
    return signer.verify_webhook_signature(payload, signature, secret, algorithm)